src/lavinmq/auth/authenticators/oauth.cr

```diff
@@ -0,0 +1,410 @@
+        {username, tags, permissions, Time.unix(expires_at)}
+      end
+
+      protected def validate_issuer(payload)
+        issuer = payload["iss"]?.try(&.as_s?)
+        raise JWT::DecodeError.new("Missing or invalid iss claim in token") unless issuer
+
+        expected = @config.oauth_issuer_url.chomp("/")
+        actual = issuer.chomp("/")
+
+        if actual != expected
+          raise JWT::DecodeError.new("Token issuer does not match the expected issuer")
+        end
+      end
+
+      protected def validate_audience(payload)
+        aud = payload["aud"]?
+        return unless aud
+
+        audiences = case aud
+                    when .as_a? then aud.as_a.map(&.as_s)
+                    when .as_s? then [aud.as_s]
+                    else             return
+                    end
+        expected = @config.oauth_audience.empty? ? @config.oauth_resource_server_id : @config.oauth_audience
+        return if expected.empty?
+
+        if !audiences.includes?(expected)
+          raise JWT::DecodeError.new("Token audience does not match expected value")
+        end
+      end
+
+      private def extract_username(payload) : String
+        @config.oauth_preferred_username_claims.each do |claim|
+          if username = payload[claim]?.try(&.as_s?)
+            return username
+          end
+        end
+        raise "No username found in JWT claims (tried: #{@config.oauth_preferred_username_claims.join(", ")})"
+      end
+
+      # Extracts roles/scopes from JWT payload and converts to LavinMQ tags and permissions.
+      private def parse_roles(payload)
+        tags = Set(Tag).new
+        permissions = Hash(String, User::Permissions).new
+        scopes = [] of String
+
+        if !@config.oauth_resource_server_id.empty?
+          if arr = payload.dig?("resource_access", @config.oauth_resource_server_id, "roles").try(&.as_a?)
+            scopes.concat(arr.map(&.as_s))
+          end
+        end
+
+        if scope_str = payload["scope"]?.try(&.as_s?)
+          scopes.concat(scope_str.split)
+        end
+
+        if !@config.oauth_additional_scopes_key.empty?
+          if claim = payload[@config.oauth_additional_scopes_key]?
+            scopes.concat(extract_scopes_from_claim(claim))
+          end
+        end
+
+        filter_scopes(scopes).each { |scope| parse_role(scope, tags, permissions) }
+        {tags.to_a, permissions}
+      end
+
+      private def extract_scopes_from_claim(claim) : Array(String)
+        case claim
+        when .as_h?
+          claim.as_h.compact_map do |key, value|
+            if @config.oauth_resource_server_id.empty? || key == @config.oauth_resource_server_id
+              extract_scopes_from_claim(value)
+            end
+          end.flatten
+        when .as_a?
+          claim.as_a.flat_map do |item|
+            extract_scopes_from_claim(item)
+          end
```

Comment:
don't need to `compact!` here?

Reply:
`extract_scopes_from_claim(item) ` is always called, no conditional here so we have no `nil` in the array?